ChocolateyGet.psm1

```diff
@@ -71,26 +73,44 @@ function Get-DynamicOptions
 }
 
 
+# Wildcard pattern matching configuration
+$script:wildcardOptions = [System.Management.Automation.WildcardOptions]::CultureInvariant -bor `
+                          [System.Management.Automation.WildcardOptions]::IgnoreCase
 
 # This function gets called during find-package, install-package, get-packagesource etc.
 # OneGet uses this method to identify which provider can handle the packages from a particular source location.
 function Resolve-PackageSource { 
 
     Write-Debug ($LocalizedData.ProviderDebugMessage -f ('Resolve-PackageSource')) 
-    
-    $isTrusted    = $false
-    $isRegistered = $false
-    $isValidated  = $true
-    $location     = $script:PackageSource
-    
- 
-    foreach($Name in @($request.PackageSources)) {
+    $SourceName = $request.PackageSources
 
-        if($Name -eq $script:PackageSourceName)
-        {
-    	    write-debug ($LocalizedData.ProviderDebugMessage -f ('Resolve-PackageSources to $location'))
+    # get Sources from the registered config file
+    [array]$RegisteredPackageSources = Get-PackageSources
+
+    if(-not $SourceName)
+    {
+        $SourceName = "*"
+    }
+
+    foreach($src in $SourceName)
+    {
+        if($request.IsCanceled) { return }
+
+        # Get the sources that registered before
+        $sourceFound = $false
+
+        $RegisteredPackageSources | Where-Object {$_.Name -like $src -and $_.Disabled -eq 'False'} |
+            ForEach-Object {
+                $packageSource = New-PackageSourceAndYield -Source $_
+                Write-Output -InputObject $packageSource
+                $sourceFound = $true
+            }
 
-            New-PackageSource $Name $location $isTrusted $isRegistered $isValidated
+        # If a user does specify -Source but not registered
+        if(-not $sourceFound)
```

Comment:
Considering your are enabling the -source, here are a few cases i think we should support:
1. find-package nodejs -provider chocolateyget      (we should make sure we do not break today's scenario, right?
2. register-packagesource -name chocolatey ...,
    find-package nodejs -provider chocolateyGet -source chocolatey
3. register-packagesource -provider chocolateyGet -source "http://somewhere.org/api/v2

Reply:
I believe case 1 still works (see tests) and case 2 already works with these changes with a default installation of chocolatey.

It continues to use the public repo by default if no source name is specified, so long as chocolatey knows about it. If someone intentionally de-registers the default public repo **then** tries to find or install a package without specifying a source, it will throw an error. Registering back the public repo restores basic functionality without having to specify a source name, though you can specify 'chocolatey' as a source if you want.

```powershell
PS C:\> get-packagesource -ProviderName chocolateyget

Name                             ProviderName     IsTrusted  Location
----                             ------------     ---------  --------
chocolatey                       ChocolateyGet    True       https://chocolatey.org/api/v2/

PS C:\> find-package nodejs -ProviderName chocolateyget

Name                           Version          Source           Summary
----                           -------          ------           -------
nodejs                         11.9.0           Chocolatey


PS C:\> Get-PackageSource -ProviderName chocolateyget  | unregister-packagesource

PS C:\> find-package nodejs -ProviderName chocolateyget
find-package : No match was found for the specified search criteria and package name 'nodejs'. Try Get-PackageSource
to see all available registered package sources.
At line:1 char:1
+ find-package nodejs -ProviderName chocolateyget
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...ets.FindPackage:FindPackage) [Find-Package], Exceptio
   n
    + FullyQualifiedErrorId : NoMatchFoundForCriteria,Microsoft.PowerShell.PackageManagement.Cmdlets.FindPackage

PS C:\> Register-PackageSource -ProviderName chocolateyget -Name chocolatey -Location 'https://chocolatey.org/api/v2/'
WARNING: Package source not found

Name                             ProviderName     IsTrusted  Location
----                             ------------     ---------  --------
chocolatey                       ChocolateyGet    True       https://chocolatey.org/api/v2/


PS C:\> find-package nodejs -ProviderName chocolateyget

Name                           Version          Source           Summary
----                           -------          ------           -------
nodejs                         11.9.0           Chocolatey
```

Case 3 works, though a little different than you describe, if you first register the package with chocolatey (either through choco.exe or Register-PackageSource). Per https://docs.microsoft.com/en-us/powershell/module/packagemanagement/find-package, the -Source argument has to take in the name of a registered package source. It cant take in an arbitrary URL that hasn't first been registered with the package provider.

Reply:
Thanks for your comments. For case #1, if a user does not run register-package source, find-package nodejs will fail, which will break today's case.  I think you may fall back to the default chocolateyget.org if a user does not register the source?

``` powershell

PS C:\WINDOWS\system32> Get-PackageSource -ProviderName chocolateyget
WARNING: Unable to find package sources.

PS C:\WINDOWS\system32> find-package nodejs -ProviderName chocolateyget -verbose
VERBOSE: Using the provider 'ChocolateyGet' for searching packages.
VERBOSE: Found choco.exe in 'C:\ProgramData\chocolatey\bin\choco.exe'.
VERBOSE: Found choco.exe in 'C:\ProgramData\chocolatey\bin\choco.exe'.
find-package : No match was found for the specified search criteria and package name 'nodejs'. Try Get-PackageSource to see all available registered package sources.
At line:1 char:1
+ find-package nodejs -ProviderName chocolateyget -verbose
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...ets.FindPackage:FindPackage) [Find-Package], Exception
    + FullyQualifiedErrorId : NoMatchFoundForCriteria,Microsoft.PowerShell.PackageManagement.Cmdlets.FindPackage
```

Reply:
Does the installation of choco on your test machine have anything listed when you run 'choco source'? A default installation of chocolatey puts chocolatey.org in c:\programdata\chocolatey\config\chocolatey.config, and the modified provider stores and retrieves information to it by calling 'choco source'.  

```powershell
PS C:\Users\WDAGUtilityAccount\Desktop\ChocolateyGet-master> choco source
Chocolatey v0.10.11
chocolatey - https://chocolatey.org/api/v2/ | Priority 0|Bypass Proxy - False|Self-Service - False|Admin Only - False.
```

The original Pester [tests](https://ci.appveyor.com/project/jianyunt/chocolateyget/builds/22091691#L352) still seem to be working, and a fresh installation of both the modified module and chocolatey (that the module installs) on a Sandbox container seems to handle finding and installing packages from chocolatey.org without any packagesource changes.

```powershell
PS C:\Users\WDAGUtilityAccount\Desktop\ChocolateyGet-master> Import-PackageProvider .\ChocolateyGet.psm1


Name                    : ChocolateyGet                                                                                 ProviderName            : ChocolateyGet                                                                                 Features                : {}                                                                                            Version                 : 0.0.0.0                                                                                       ProviderPath            : C:\Users\WDAGUtilityAccount\Desktop\ChocolateyGet-master\ChocolateyGet.psm1                   SupportedFileExtensions : {}
SupportedUriSchemes     : {}
DynamicOptions          : {AdditionalArguments, AdditionalArguments}
FastPackageReference    :
Source                  :
Status                  :
SearchKey               :
FullPath                :
PackageFilename         :
FromTrustedSource       : False
Summary                 :
SwidTags                : {ChocolateyGet}
CanonicalId             :
Metadata                : {}
SwidTagText             : <?xml version="1.0" encoding="utf-16" standalone="yes"?>
                          <SoftwareIdentity
                            version="0.0.0.0"
                            name="ChocolateyGet" xmlns="http://standards.iso.org/iso/19770/-2/2015/schema.xsd" />
Dependencies            : {}
IsCorpus                :
VersionScheme           :
TagVersion              :
TagId                   :
IsPatch                 :
IsSupplemental          :
AppliesToMedia          :
Meta                    : {}
Links                   : {}
Entities                : {}
Payload                 :
Evidence                :
Culture                 :
Attributes              : {version,name}



PS C:\Users\WDAGUtilityAccount\Desktop\ChocolateyGet-master> Find-Package nodejs -ProviderName chocolateyget

Choco.exe is required to continue
ChocolateyGet is built on Choco.exe. Do you want ChocolateyGet to install Choco.exe from
'https://chocolatey.org/install.ps1' now?
[Y] Yes  [N] No  [S] Suspend  [?] Help (default is "Y"): y
WARNING: Cannot bind argument to parameter 'message' because it is an empty string.


FastPackageReference : nodejs#11.9.0#Chocolatey
ProviderName         : ChocolateyGet
Source               : Chocolatey
Status               : Available
SearchKey            :
FullPath             :
PackageFilename      :
FromTrustedSource    : True
Summary              :
SwidTags             : {nodejs}
CanonicalId          : chocolateyget:nodejs/11.9.0#Chocolatey
Metadata             : {}
SwidTagText          : <?xml version="1.0" encoding="utf-16" standalone="yes"?>
                       <SoftwareIdentity
                         name="nodejs"
                         version="11.9.0"
                         versionScheme="MultiPartNumeric"
                       xmlns="http://standards.iso.org/iso/19770/-2/2015/schema.xsd">
                         <Meta />
                       </SoftwareIdentity>
Dependencies         : {}
IsCorpus             :
Name                 : nodejs
Version              : 11.9.0
VersionScheme        : MultiPartNumeric
TagVersion           :
TagId                :
IsPatch              :
IsSupplemental       :
AppliesToMedia       :
Meta                 : {{}}
Links                : {}
Entities             : {}
Payload              :
Evidence             :
Culture              :
Attributes           : {name,version,versionScheme}
```

Reply:
Just checking in to see if you've had further chance to review this. 

Wondering if perhaps the issue you saw in your last comment was due to unregistering the default chocolatey.org repo by running the following from my previous reply:

```powershell
PS C:\> Get-PackageSource -ProviderName chocolateyget  | unregister-packagesource
```